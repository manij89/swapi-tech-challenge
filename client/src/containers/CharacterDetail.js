/* eslint-disable react-hooks/exhaustive-deps */
import React, { useContext, useEffect, useState } from 'react';
import {
  Center,
  Image,
  Box,
  Link,
  Badge,
  UnorderedList,
  ListItem,
  Text,
  Button,
  useColorMode,
} from '@chakra-ui/react';
import ls from 'local-storage';
import Spinner from '../components/Spinner';
import { useHistory, Link as ReachLink } from 'react-router-dom';
import { Context } from '../context/Store';
import { bagdeBg, bgColor, textColor } from '../styles/colorModes';
import { handleSaveCharacter } from '../helpers/utils';

export default function Characterdetail({ name }) {
  const [state, dispatch] = useContext(Context);
  const [charFilms, setCharFilms] = useState([]);
  const history = useHistory();
  const { colorMode } = useColorMode();

  const getCharMovies = () => {
    const list = [...state.allFilms] || ls.get('allFilms');
    const result = list.filter(f => state.clickedChar.films.includes(f.url));
    setCharFilms(result);
  };

  const handleClick = film => {
    import(`../images/${film.episode_id}.jpg`).then(image =>
      dispatch({
        type: 'SET_CLICKED_FILM',
        payload: { ...film, poster: image.default },
      })
    );
  };

  useEffect(() => {
    dispatch({ type: 'SET_LOADING', payload: true });
    if (state.clickedChar.name) {
      getCharMovies();
    } else {
      history.goBack();
    }
    dispatch({ type: 'SET_LOADING', payload: false });
  }, [state.clickedChar]);

  const {
    image,
    wiki,
    height,
    hair_color,
    skin_color,
    birth_year,
  } = state.clickedChar;
  return (
    <>
      {state.loading ? (
        <Spinner />
      ) : (
        <Center
          flexDirection={['column', 'column', 'row', 'row']}
          mx="10px"
          my={['10px', '15px', '40px', '40px']}
          rounded="md"
          overflow="hidden"
          boxShadow="lg"
          bg={bgColor[colorMode]}
        >
          <Image
            m="10px"
            src={image}
            alt={name}
            h="auto"
            w={['60%', '50%', '50%', '50%']}
          />
          <Box p={5} w="100%">
            <Text
              as="h2"
              fontWeight="bold"
              fontSize="1.5rem"
              my={2}
              color={textColor[colorMode]}
            >
              {name}
            </Text>
            <UnorderedList listStyleType="none">
              <ListItem>Height: {height}</ListItem>
              <ListItem>Hair Color: {hair_color}</ListItem>
              <ListItem>Skin: {skin_color}</ListItem>
              <ListItem>Birth year: {birth_year}</ListItem>
              <ListItem>
                More info: <Link>{wiki}</Link>
              </ListItem>
            </UnorderedList>

            <Box my={2}>
              {charFilms.map(f => (
                <Link as={ReachLink} to={`/film/${f.episode_id}`}>
                  <Badge
                    key={f.title}
                    variant="subtle"
                    rounded="full"
                    fontSize={13}
                    m={1}
                    p={1}
                    _hover={{ boxShadow: 'md', cursor: 'pointer' }}
                    _active={{ boxShadow: 'lg' }}
                    onClick={() => handleClick(f)}
                  >
                    {f.title}
                  </Badge>
                </Link>
              ))}
            </Box>
            <Box textAlign="center">
              <Button
                size="lg"
                mt={3}
                color="white"
                bg={bagdeBg[colorMode]}
                boxShadow="sm"
                onClick={() =>
                  handleSaveCharacter(
                    state.clickedChar,
                    state.favoriteCharacters,
                    dispatch
                  )
                }
                _hover={{ boxShadow: 'md' }}
                _active={{ boxShadow: 'lg' }}
              >
                {state.favoriteCharacters.find(char => char.name === name)
                  ? 'Saved'
                  : 'Add to Favorites'}
              </Button>
            </Box>
          </Box>
        </Center>
      )}
    </>
  );
}
