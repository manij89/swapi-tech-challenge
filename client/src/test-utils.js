import React from 'react';
import { render } from '@testing-library/react';
import { Route, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { ChakraProvider, theme } from '@chakra-ui/react';
import Store from './context/Store';

const AllProviders = ({ children }) => (
  <Store>
    <Router history={history}>
      <ChakraProvider theme={theme}>{children}</ChakraProvider>
    </Router>
  </Store>
);

const customRender = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    initialState,
    store,
    ...options
  } = {}
) => {
  const Wrapper = ({ children }) => (
    <Store>
    <Router history={history}>
      <ChakraProvider theme={theme}>{children}</ChakraProvider>
    </Router>
  </Store>
  );

  return {
    ...render(ui, {
      wrapper: Wrapper,
      ...options,
    }),
    history,
    store,
  };
};

export { customRender as render };
