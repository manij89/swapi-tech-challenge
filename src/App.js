import React from 'react';
import {
  ChakraProvider,
  theme,
} from '@chakra-ui/react';
import {  Route, Switch } from 'react-router-dom';
import Home from './containers/Home';
import FavFilms from './containers/FavFilms';
import FavCharacter from './containers/FavCharacter';


function App() {
  return (
    <ChakraProvider theme={theme}>
      <Switch>
        <Route exact path='/' component={Home} />
        <Route exact path='/favFilms' component={FavFilms} />
        <Route exact path='/favCharacter' component={FavCharacter} />
      </Switch>

    </ChakraProvider>
  );
}

export default App;
