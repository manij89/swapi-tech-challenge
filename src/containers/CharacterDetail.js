import React, { useContext, useEffect, useRef } from 'react';
import {
  Center,
  Image,
  Box,
  Stack,
  Badge,
  Text,
  Button,
  useColorMode,
} from '@chakra-ui/react';
import { useHistory } from 'react-router-dom';
import { Context } from '../context/Store';
import { bgColor, textColor, bagdeBg } from '../styles/colorModes';
import { handleSaveCharacter } from '../helpers/utils';

export default function Characterdetail({ name }) {
  const [state, dispatch] = useContext(Context);
  const history = useHistory();
  const { colorMode } = useColorMode();
  const currChar = useRef(
    state.characterDetails.find(char => char.name === name)
  );
  const extraInfoChar = useRef(
    state.allImages.find(char => char.name === name)
  );

  const getCharInfo = () => {
    if (state.characterDetails.length && state.allImages.length) {
      dispatch({
        type: 'SET_CLICKED_CHAR',
        payload: {
          ...currChar.current,
          image: extraInfoChar.image,
          wiki: extraInfoChar.wiki,
        },
      });
    } else {
      history.goBack();
    }
  };

  useEffect(() => {
    dispatch({ type: 'SET_LOADING', payload: true });
    getCharInfo();
    dispatch({ type: 'SET_LOADING', payload: false });
  }, []);

  return (
    <>
      {state.loading ? (
        'loading...'
      ) : (
        <Center
          flexDirection={['column', 'column', 'row', 'row']}
          mx="10px"
          my={['10px', '15px', '40px', '40px']}
          rounded="md"
          overflow="hidden"
          boxShadow="lg"
          bg={bgColor[colorMode]}
        >
          <Image
            m="10px"
            src={extraInfoChar.current.image}
            alt={name}
            h="auto"
            w={['60%', '50%', '50%', '50%']}
          />
          <Box p={5} w="100%">
            <Text
              as="h2"
              fontWeight="bold"
              fontSize="1.5rem"
              my={2}
              color={textColor[colorMode]}
            >
              {name}
            </Text>
            {/* <Text color={textColor[colorMode]} fontWeight="light" fontSize="md">
            {opening_crawl}
          </Text> */}
            {/* <Box my={2}>
            {characters.map(char => (
              <Badge
                key={char.name}
                variant="subtle"
                rounded="full"
                fontSize={13}
                m={1}
                p={1}
                _hover={{ boxShadow: 'md', cursor: 'pointer' }}
                _active={{ boxShadow: 'lg' }}
                onClick={() => history.push(`/character/${char.name}`)}
              >
                {char.name}
              </Badge>
            ))}
          </Box> */}
            <Box textAlign="center">
              <Button
                size="lg"
                mt={3}
                boxShadow="sm"
                onClick={() =>
                  handleSaveCharacter(
                    state.clickedChar,
                    state.favoriteCharacters,
                    dispatch
                  )
                }
                _hover={{ boxShadow: 'md' }}
                _active={{ boxShadow: 'lg' }}
              >
                {state.favoriteCharacters.find(char => char.name === name)
                  ? 'Saved'
                  : 'Add to Favorites'}
              </Button>
            </Box>
          </Box>
        </Center>
      )}
    </>
  );
}
