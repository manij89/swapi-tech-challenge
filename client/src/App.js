/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useContext } from 'react';
import { Context } from './context/Store';
import * as apiClient from './helpers/apiClient';
import { ChakraProvider } from '@chakra-ui/react';
import { customTheme } from './styles/colorModes';
import { Route, Switch } from 'react-router-dom';
import Home from './containers/Home';
import FavFilms from './containers/FavFilms';
import FavCharacter from './containers/FavCharacter';
import FilmDetail from './containers/FilmDetail';
import CharacterDetail from './containers/CharacterDetail';
import Navbar from './components/Navbar/NavBar';

function App() {
  const [_, dispatch] = useContext(Context);

  useEffect(() => {
    apiClient.getMovies(dispatch);
    apiClient.getCharacterImages(dispatch);
  }, []);

  return (
    <ChakraProvider theme={customTheme}>
      <Navbar />
      <Switch>
        <Route exact path="/" component={Home} />
        <Route
          exact
          path="/film/:id"
          render={routeProps => (
            <>
              <FilmDetail episode_id={routeProps.match.params.id} />
            </>
          )}
        />
        <Route
          exact
          path="/character/:name"
          render={routeProps => (
            <>
              <CharacterDetail name={routeProps.match.params.name} />
            </>
          )}
        />
        <Route exact path="/favFilms" component={FavFilms} />
        <Route exact path="/favCharacter" component={FavCharacter} />
      </Switch>
    </ChakraProvider>
  );
}

export default App;
