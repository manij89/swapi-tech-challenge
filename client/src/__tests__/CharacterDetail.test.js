import React from 'react';
import { cleanup} from '@testing-library/react'
import { createMemoryHistory } from 'history';
import App from '../App';
import CharacterDetail from '../containers/CharacterDetail';

import {render} from '../test-utils';

afterEach(cleanup)

it('register snapshot', () => {
  const history = createMemoryHistory();
  history.push('/character/Yoda');
  const { asFragment } = render(<App />, { history });
  
  expect(asFragment(<CharacterDetail />)).toMatchSnapshot()
})