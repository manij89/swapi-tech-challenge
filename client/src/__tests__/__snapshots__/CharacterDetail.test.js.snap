// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-5ysws4"
  >
    <div
      class="css-rkk3mp"
    >
      <div
        class="css-k008qs"
      >
        <div
          class="css-0"
          id="container"
        >
          <div
            class="false css-0"
            id="burger"
          >
            <div
              class="bun top css-0"
            />
            <div
              class="filling css-0"
            />
            <div
              class="bun bottom css-0"
            />
          </div>
        </div>
        <nav
          class="nav css-lnyvlk"
        >
          <ul
            class="chakra-stack css-uqqhq"
          >
            <li
              class="green css-0"
            >
              <a
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="red css-0"
            >
              <a
                href="/favFilms"
              >
                Favorite Films
              </a>
            </li>
            <li
              class="blue css-0"
            >
              <a
                href="/favcharacter"
              >
                Favorite Characters
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
    <p
      class="chakra-text css-19n1896"
    >
      STAR WARS
    </p>
    <img
      aria-label="Switch to light mode"
      class="chakra-image__placeholder css-d25thr"
    />
  </div>
  <div
    class="css-lr27e3"
    data-testid="list-container"
  />
</DocumentFragment>
`;
